### PR TITLE
[AMD-ZENDNN] Code changes to support TF-Plugin for AMD CPU Inference

### PR DESCRIPTION
**Pull request by aadwived**
This pull request contains necessary/base changes for supporting TF-Plugin for AMD CPU Inference.

With this code change we introduce,
    - Functionality to enable AMD CPU optimizations by setting environment variable 'TF_ENABLE_ZENDNN_OPTS=1'
    - Zen layout pass to re-write supported ops with ZenOps

All these code changes are under AMD_ZENDNN flag which is enabled for linux builds only.

Authors:
    Aakar Dwivedi ( aakar.dwivedi@amd.com )
    Aditya Chatterjee ( aditya.chatterjee@amd.com )
    Arun Coimbatore Ramachandran ( aruncoimbatore.ramachandran@amd.com )
    Avinash-Chandra Pandey ( avinash-chandra.pandey@amd.com )
    Chandra Kumar Ramasamy ( chandrakumar.ramasamy@amd.com )
    Savan Anadani ( savan.anadani@amd.com )

Signed-off-by: Aakar Dwivedi <Aakar.Dwivedi@amd.com>

### CODE REVIEW

File: tensorflow/core/util/BUILD [hunk 1]
```diff
@@ -162,6 +162,7 @@ filegroup(
         "mkl_threadpool.h",
         "mkl_util.h",
         "onednn_env_vars.h",
+        "zen_util.h",
```
Comment (penpornk) [inline tensorflow/core/util/BUILD:165]: Please guard this with `if_zendnn`.

File: tensorflow/core/util/zen_util.h [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+#include <sstream>
+#include <string>
+
+namespace tensorflow {
+
+/// Read an integer from the environment variable
+/// Return default_value if the environment variable is not defined, otherwise
+/// return actual value.
+inline int zendnn_getenv_int(const char *name, int default_value = 0) {
+  char *val = std::getenv(name);
```
Comment (penpornk) [inline tensorflow/core/util/zen_util.h:38]: Syscalls are expensive. We want to read the environment variable only once and cache the results for subsequent calls. Please follow the approach in `IsZenDnnDisabled()`.

File: tensorflow/core/util/port.h [hunk 3]
```diff
@@ -44,6 +44,9 @@ bool GpuSupportsHalfMatMulAndConv();
 // Returns true if INTEL_MKL is defined
 bool IsMklEnabled();
 
+// Returns true if TF_ENABLE_ZENDNN_OPTS is set to 0
+bool IsZenDNNDisabled();
```
Comment (penpornk) [inline tensorflow/core/util/port.h:48]: Nit: Camel case.
```suggestion
bool IsZenDnnDisabled();
```

File: tensorflow/core/util/zen_util.h [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+  return val == NULL ? default_value : atoi(val);
+}
+
+#if !defined(LOG_LEVEL_DEFAULT)
+#define LOG_LEVEL_DEFAULT LOG_LEVEL_ERROR
+#endif
+
+enum LogLevel {
```
Comment (penpornk) [inline tensorflow/core/util/zen_util.h:46]: May I ask why we can't use TF's existing logging framework?

File: tensorflow/core/common_runtime/BUILD [hunk 5]
```diff
@@ -1844,6 +1862,7 @@ tf_cuda_library(
         ":memory_types",
         ":mkl_cpu_allocator",
         ":mkl_layout_pass",
+        ":zen_layout_pass",
```
Comment (penpornk) [inline tensorflow/core/common_runtime/BUILD:1865]: Please wrap this with `if_zendnn`.

File: tensorflow/core/common_runtime/BUILD [hunk 6]
```diff
@@ -1251,6 +1251,24 @@ cc_library(
     alwayslink = 1,
 )
 
+cc_library(
+    name = "zen_layout_pass",
+    srcs = ["zen_layout_pass.cc"],
+    copts = tf_copts() + ["-fopenmp"],
```
Comment (penpornk) [inline tensorflow/core/common_runtime/BUILD:1257]: Vanilla TF shouldn't use OpenMP. I don't think `zen_layout_pass` needs OpenMP so we can just remove it.
```suggestion
    copts = tf_copts(),
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for your PR and I'm so sorry for the late review!

**Comment by gbaned** [comment]
Hi @aadwived Can you please check @penpornk's comments and resolve conflicts?. Thank you!

**Pull request closed without merging**